File: 골드 3 (1문제)/16964_DFS 스페셜 저지.py
# 설명 링크 : https://moon-gd.tistory.com/98
import sys
q_input = sys.stdin.readline  # 빠른 입력

size = int(q_input())  # 크기 입력

# 그래프 리스트 생성
graph = [[]]
for _ in range(size):
    graph.append([])

visited = [False for _ in range(size + 1)]  # 방문 리스트 생성
depth = [0 for _ in range(size + 1)]  # 트리 깊이 리스트 생성
sub_node = [0 for _ in range(size + 1)]

for _ in range(size - 1):
    v1, v2 = map(int, q_input().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

input_order = list(map(int, q_input().split()))
input_order.insert(0, 0)
if input_order[1] != 1:
    print(0)
    exit(0)


def dfs(vertex, level):
    if visited[vertex]:
        return 0
    else:
        visited[vertex] = True  # 방문 표시
        depth[vertex] = level  # 트리 깊이 저장

        child = 1
        for v in graph[vertex]:
            # 방문하지 않은 정점인 경우
            if not visited[v]:
                # DFS 탐색
                child += dfs(v, level + 1)

        sub_node[vertex] = child

        return child


dfs(1, 0)

flag = 1
for i in range(2, size + 1):
    if sub_node[input_order[i]] + i > size:
        continue
    else:
        next_vertex = input_order[sub_node[input_order[i]] + i]
        if depth[next_vertex] > depth[input_order[i]]:
            flag = 0
            break


print(flag)

File: 골드 3 (1문제)/16964_DFS 스페셜 저지.c
// 설명 링크 : https://moon-gd.tistory.com/98
#define _CRT_SECURE_NO_WARNINGS

#include <stdio.h>
#include <stdlib.h>
#define MAX 100001

typedef struct node* nodePointer;
typedef struct node {
	int value;
	nodePointer next;
}node;
node graph[MAX] = { 0, NULL };

int visited[MAX];
int input_order[MAX];
int depth[MAX];
int child[MAX];

void insert(int start, int end);
int dfs(int start, int count);

int main() {
	int size = 0;
	int v1, v2; // 정점 임시 변수

	// 입력
	scanf("%d", &size);
	for (int i = 1; i < size; i++) {
		scanf("%d %d", &v1, &v2);
		insert(v1, v2);
		insert(v2, v1);
	}

	dfs(1, 0);

	for (int i = 1; i <= size; i++) {
		scanf("%d", &input_order[i]);
	}

	if (input_order[1] != 1) { printf("0"); return 0; }

	for (int i = 2; i <= size; i++) {
		if (i + child[input_order[i]] > size) {
			continue;
		}

		int next_depth = depth[input_order[i + child[input_order[i]]]];

		if (next_depth > depth[input_order[i]]) {
			printf("0");
			return 0;
		}
	}

	printf("1");

	return 0;
}

void insert(int start, int end) {
	nodePointer temp = (nodePointer)malloc(sizeof(node));
	temp->next = NULL;
	temp->value = end;

	if (graph[start].next) {
		temp->next = graph[start].next;
		graph[start].next = temp;
	}
	else {
		graph[start].next = temp;
	}
}

int dfs(int start, int count) {
	if (visited[start] == 1) {
		return 0;
	}
	else {
		visited[start] = 1;
		depth[start] = count;
		nodePointer searcher = graph[start].next;
		int child_count = 1;

		for (; searcher; searcher = searcher->next) {
			if (visited[searcher->value] == 0) {
				child_count += dfs(searcher->value, count + 1);
			}
		}

		return child[start] = child_count;
	}
}